text/0000-list-rendering-reform.md

```diff
@@ -0,0 +1,306 @@
+  <button key={item.id}></button>
+</template>
+```
+
+With the new directives, this becomes:
+
+```html
+<template lwc:each={items} lwc:item="item" lwc:key={item.id}>
+  <div></div>
+  <span></span>
+  <button></button>
+</template>
+```
+
+Also note that, in the shorthand `<template>`-less format, `lwc:key` is defined not on a `<template>` but on the same element as the `lwc:each`:
+
+
+```html
+<li lwc:each={items} lwc:item="item" lwc:key={item.id}>
+    {item.name}
+</li>
+```
+
+Similar to `key`, `lwc:key` is required on elements with `lwc:each`, and throws an error if missing:
+
+    Missing lwc:key for lwc:each. Iterators must have a unique, computed key value.
+
+Unlike `key`, a missing `lwc:key` throws the error regardless of the content inside the iterator. (A missing `key` will not throw for text nodes, comment nodes, or empty markup – an apparent [inconsistency](https://github.com/salesforce/lwc/issues/3860) in that directive.)
+
+Similar to `key`, `lwc:key` must not reference the variable defined by `lwc:index`. It also must not reference the variable defined by `lwc:first`, `lwc:last`, `lwc:even`, or `lwc:odd`. For example, this will throw a compile-time error:
+
+```html
+<template lwc:each={items} lwc:item="item" lwc:index="idx" lwc:key={idx}>
+</template>
+```
+
+In short, `lwc:key` _must_ reference the `lwc:item` variable. Anything else throws a compile-time error.
+
+### `lwc:index`
+
+Identical to `for:index`. Returns a number for the current list index.
+
+As with `for:index`, `lwc:index` cannot use the same string as the `lwc:item` on the same element:
+
+```html
+<template lwc:each={items} lwc:key={same.id} lwc:item="same" lwc:index="same">
+</template>
+```
+
+The above will throw a compile-time error.
+
+(Note that `for:index` only [incidentally](https://stackblitz.com/edit/salesforce-lwc-se51rr?file=src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fapp%2Fapp.js&title=LWC%20playground) throws an error for the same case, because the `key` ends up referencing the `for:index` variable. For `lwc:index`, the compile-time error should be explicit.)
+
+### `lwc:first`, `lwc:last`, `lwc:even`, `lwc:odd`
+
+These new directives take strings as their values. When referenced, they resolve to a boolean representing whether the list item is first, last, even, or odd. (This is heavily inspired by [Angular](https://angular.io/api/common/NgForOf#local-variables).)
+
+The addition of `lwc:first` and `lwc:last` serve to unify the two existing directives – previously, `iterator:*` was the only way to get "first" or "last" behavior. The addition of `lwc:even` and `lwc:odd` are merely for convenience (especially in the absence of [complex template expressions](https://github.com/salesforce/lwc/pull/3376)).
+
+### Constraints
+
+#### Variable reuse
+
+All new directives that take a string as their value (`lwc:item`, `lwc:index`, `lwc:first`, `lwc:last`, `lwc:even`, and `lwc:odd`) must have unique values when used on the same element.

```

Comment:
Any thoughts on using `{}` to wrap the `lwc:item` instead of `""`?  It's something that I have always found inconsistent in the current syntax. 

In LWC template syntax, the usage of quotes on an HTML attribute denotes a static binding. But in the case of `lwc:item`, it is used to indicate a dynamic binding to a variable. 

For context, I was the one who originally pushed to wrap the `lwc:item` value within quotes to discourage users from using JavaScript expressions. That said, my view changed on this over the years. I would like to get the sentiment from the rest of the team on this.

Reply:
I think using `""` would make the syntax more consistent with [scoped slots](https://help.salesforce.com/s/articleView?id=release-notes.rn_lwc_scoped_slots.htm&release=244&type=5):

```html
    <c-child>
        <template lwc:slot-data="item">
            <span>{item.name}</span>
        </template>
    </c-child>
```

In my mind, the quotes mean "declare a variable with a given name," not "evaluate a JS expression." Also, if we change `lwc:item`, we would arguably want to change `lwc:index`, `lwc:first`, `lwc:last`, `lwc:even`, and `lwc:odd` as well, since they also declare a variable. For those reasons, I'm in favor of keeping `""`.